File: code/rust/ch08/src/state/lua_table.rs
use super::lua_value::LuaValue;
use std::collections::HashMap;
use std::hash::{Hash, Hasher};

#[derive(Clone)]
pub struct LuaTable {
    arr: Vec<LuaValue>,
    map: HashMap<LuaValue, LuaValue>,
    rdm: usize, // hash code
}

impl Hash for LuaTable {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.rdm.hash(state);
    }
}

impl LuaTable {
    pub fn new(narr: usize, nrec: usize) -> LuaTable {
        LuaTable {
            arr: Vec::with_capacity(narr),
            map: HashMap::with_capacity(nrec),
            rdm: super::math::random(),
        }
    }

    pub fn len(&self) -> usize {
        self.arr.len()
    }

    pub fn get(&self, key: &LuaValue) -> LuaValue {
        if let Some(idx) = to_index(key) {
            if idx <= self.arr.len() {
                return self.arr[idx - 1].clone(); // TODO
            }
        }
        if let Some(val) = self.map.get(key) {
            val.clone() // TODO
        } else {
            LuaValue::Nil
        }
    }

    pub fn put(&mut self, key: LuaValue, val: LuaValue) {
        if key.is_nil() {
            panic!("table index is nil!");
        }
        if let LuaValue::Number(n) = key {
            if n.is_nan() {
                panic!("table index is NaN!");
            }
        }

        if let Some(idx) = to_index(&key) {
            let arr_len = self.arr.len();
            if idx <= arr_len {
                let val_is_nil = val.is_nil();
                self.arr[idx - 1] = val;
                if idx == arr_len && val_is_nil {
                    self.shrink_array();
                }
                return;
            }
            if idx == arr_len + 1 {
                self.map.remove(&key);
                if !val.is_nil() {
                    self.arr.push(val);
                    self.expand_array();
                }
                return;
            }
        }

        if !val.is_nil() {
            self.map.insert(key, val);
        } else {
            self.map.remove(&key);
        }
    }

    fn shrink_array(&mut self) {
        while !self.arr.is_empty() {
            if self.arr.last().unwrap().is_nil() {
                self.arr.pop();
            } else {
                break;
            }
        }
    }

    fn expand_array(&mut self) {
        let mut idx = self.arr.len() + 1;
        loop {
            let key = LuaValue::Integer(idx as i64);
            if self.map.contains_key(&key) {
                let val = self.map.remove(&key).unwrap();
                self.arr.push(val);
                idx += 1;
            } else {
                break;
            }
        }
    }
}

fn to_index(key: &LuaValue) -> Option<usize> {
    if let LuaValue::Integer(i) = key {
        if *i >= 1 {
            return Some(*i as usize);
        }
    } else if let LuaValue::Number(n) = key {
        if let Some(i) = super::math::float_to_integer(*n) {
            if i >= 1 {
                return Some(i as usize);
            }
        }
    }
    None
}

File: code/csharp/ch10/luavm/state/LuaTable.cs
using System;
using System.Collections.Generic;
using System.Linq;
using LuaType = System.Int32;
using Math = luavm.number.Math;

namespace luavm.state
{
    public struct LuaTable
    {
        private object[] arr;
        private Dictionary<object, object> _map;

        public static LuaTable newLuaTable(int nArr, int nRec)
        {
            var t = new LuaTable();
            if (nArr > 0)
            {
                t.arr = new object[nArr];
            }

            if (nRec > 0)
            {
                t._map = new Dictionary<object, object>(nRec);
            }

            return t;
        }

        public object get(object key)
        {
            key = _floatToInteger(key);
            if (LuaValue.isInteger(key))
            {
                if (arr != null)
                {
                    var idx = LuaValue.toInteger(key);
                    if (idx >= 1 && idx <= arr.Length)
                    {
                        return arr[idx - 1];
                    }
                }
            }

            return _map.ContainsKey(key) ? _map[key] : null;
        }

        private static object _floatToInteger(object key)
        {
            if (LuaValue.isFloat(key))
            {
                var f = LuaValue.toFloat(key);
                return Math.FloatToInteger(f).Item1;
            }

            return key;
        }

        private void _shrinkArray()
        {
            for (var i = arr.Length - 1; i >= 0; i--)
            {
                if (arr[i] == null)
                {
                    Array.Copy(arr, 0, arr, 0, i);
                }
            }
        }

        private void _expandArray()
        {
            for (var idx = arr.Length + 1; true; idx++)
            {
                if (_map != null && _map.ContainsKey(idx))
                {
                    var val = _map.Values.ElementAt(idx);
                    _map.Remove(idx);
                    var b = arr.ToList();
                    b.Add(val);
                    arr = b.ToArray();
                }
                else
                {
                    break;
                }
            }
        }

        public int len()
        {
            return arr.Length;
        }

        public void put(object key, object val)
        {
            if (key == null)
            {
                throw new Exception("table index is nil!");
            }

            if (LuaValue.isFloat(key) && double.IsNaN(LuaValue.toFloat(key)))
            {
                throw new Exception("table index is NaN!");
            }

            key = _floatToInteger(key);

            if (LuaValue.isInteger(key))
            {
                var idx = LuaValue.toInteger(key);
                if (idx >= 1)
                {
                    var arrLen = arr?.Length ?? -1;
                    if (idx <= arrLen)
                    {
                        if (arr != null) arr[idx - 1] = val;
                        if (idx == arrLen && val == null)
                        {
                            _shrinkArray();
                        }

                        return;
                    }

                    if (idx == arrLen + 1)
                    {
                        _map?.Remove(idx);

                        if (val != null)
                        {
                            if (arr == null)
                            {
                                var b = new List<object> {val};
                                arr = b.ToArray();
                                _expandArray();
                            }
                            else
                            {
                                var b = arr.ToList();
                                b.Add(val);
                                arr = b.ToArray();
                                _expandArray();
                            }
                        }

                        return;
                    }
                }
            }

            if (val != null)
            {
                if (_map == null)
                {
                    _map = new Dictionary<object, object>(8);
                }

                if (!_map.ContainsKey(key))
                {
                    _map.Add(key, val);
                }
                else
                {
                    _map[key] = val;
                }
            }
            else
            {
                _map.Remove(key);
            }
        }
    }
}